MoreLinq/ToArrayByIndex.cs

```diff
@@ -0,0 +1,253 @@
+        /// sequence. The size of the array will be as large as the highest
+        /// index returned by the <paramref name="indexSelector"/> plus 1.
+        /// </returns>
+        /// <remarks>
+        /// This method forces immediate query evaluation. It should not be
+        /// used on infinite sequences. If more than one element maps to the
+        /// same index then the latter element overwrites the former in the
+        /// resulting array.
+        /// </remarks>
+
+        public static TResult[] ToArrayByIndex<T, TResult>(this IEnumerable<T> source,
+            Func<T, int> indexSelector, Func<T, TResult> resultSelector)
+        {
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+            return source.ToArrayByIndex(indexSelector, (e, _) => resultSelector(e));
+        }
+
+        /// <summary>
+        /// Creates an array from an <see cref="IEnumerable{T}"/> where a
+        /// function is used to determine the index at which an element will
+        /// be placed in the array. The elements are projected into the array
+        /// via an additional function.
+        /// </summary>
+        /// <param name="source">The source sequence for the array.</param>
+        /// <param name="indexSelector">
+        /// A function that maps an element to its index.</param>
+        /// <param name="resultSelector">
+        /// A function to project a source element into an element of the
+        /// resulting array.</param>
+        /// <typeparam name="T">
+        /// The type of the element in <paramref name="source"/>.</typeparam>
+        /// <typeparam name="TResult">
+        /// The type of the element in the resulting array.</typeparam>
+        /// <returns>
+        /// An array that contains the projected elements from the input
+        /// sequence. The size of the array will be as large as the highest
+        /// index returned by the <paramref name="indexSelector"/> plus 1.
+        /// </returns>
+        /// <remarks>
+        /// This method forces immediate query evaluation. It should not be
+        /// used on infinite sequences. If more than one element maps to the
+        /// same index then the latter element overwrites the former in the
+        /// resulting array.
+        /// </remarks>
+
+        public static TResult[] ToArrayByIndex<T, TResult>(this IEnumerable<T> source,
+            Func<T, int> indexSelector, Func<T, int, TResult> resultSelector)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (indexSelector == null) throw new ArgumentNullException(nameof(indexSelector));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            var lastIndex = -1;
+            var indexed = (List<KeyValuePair<int, T>>) null;
+            List<KeyValuePair<int, T>> Indexed() => indexed ?? (indexed = new List<KeyValuePair<int, T>>());
```

Comment:
I'd stick with `Indexed` because the accessibility of a method in .NET does not conventionally change its casing (i.e. Pascal) and `Indexed` here is certainly not a variable like `indexed` in the line immediately above. It seems simplest to me to just continue to apply the convention for methods to local functions (which are not functions really but methods as that is how they are hoisted).

> It took me a while to find this Indexed function ;)

I can imagine, especially if you're reading code online and not in an IDE where you can use a go-to-definition-like feature.